730 retinal oct rpd segmentation

ybagdasa opened the pull request:
Fixes # .

### Description
This is a proposed segmentation model to add to the model zoo.

### Status
Ready for review/work in progress.

### Please ensure all the checkboxes:
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [ ] In-line docstrings updated.
- [ ] Update `version` and `changelog` in `metadata.json` if changing an existing bundle.
- [ x] Please ensure the naming rules in config files meet our requirements (please refer to: `CONTRIBUTING.md`).
- [x ] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `metadata.json`.
- [x ] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [x ] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.
- [ x] Avoid using path that contains personal information within config files (such as use `/home/your_name/` for `"bundle_root"`).

ybagdasa committed:
First commit for forked branch.

ybagdasa committed:
Deleted the zip file download in inference.py. Updated urls in large_files.yaml. Fixed case-sensitivity for .png/.PNG in dataset/data.py. Made confidence threshold an external argument.

ybagdasa committed:
Implemented unit testing and added sample_data.

ybagdasa committed:
Applied changes from isort, black, flake in runtests.sh. Ran pytype separately.

ybagdasa committed:
Removed configargparser and added PNG naming instructions in readme.

ybagdasa committed:
Merge branch 'dev' into 730-retinalOCT_RPD_segmentation. Upstream
updates.

ybagdasa committed:
Moved large_files.yaml into retinalOCT_RPD_Segmentation into bundle root.

ybagdasa committed:
Updated dependencies. Updated depreciated pandas api. Updated install instructions in readme.

ybagdasa committed:
Merge from upstream.

ybagdasa commented:
@yiheng-wang-nv here is my draft PR

yiheng-wang-nv commented on ci/install_scripts/install_retinalOCT_RPD_segmentation_dependency.sh:
```diff
@@ -0,0 +1,3 @@
+# ensure that using evironment with python==3.9
+pip install -r ./docs/requirements.txt #includes monai
+python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'
```
Hi @ybagdasa , could you fix the commit for the `detectron2` installation?

ybagdasa replied:
@yiheng-wang-nv Thanks for considering the contribution. For the detectron2 installation, what is the command I should run to replicate the error you're getting? I know that the install works for python<3.10. Could this be the source of the issue for the ci install checks?

yiheng-wang-nv replied:
Hi @ybagdasa , 

For the comment of this line (`python -m pip install 'git+https://github.com/facebookresearch/detectron2.git'`),
based on https://github.com/Project-MONAI/model-zoo/pull/748/files#diff-6d26dc7c0d9fa894d66c944bd9d0a7c970f95371163c70631b8d6b0ed790fc81R19 `detectron2 0.6` is used, thus here we can modify the install command into:
```
python -m pip install 'git+https://github.com/facebookresearch/detectron2.git@v0.6'
```
Otherwise, the command will install the latest commit, which may have incompatible issues in the future.

yiheng-wang-nv replied:
For the python version, can you confirm if python 3.9 is necessary? 
I'm not 100% sure, but based on https://detectron2.readthedocs.io/en/latest/tutorials/install.html#requirements , it only requires to use python >= 3.7.

Currently, our repo uses python 3.10 in default for CI tests. If we do need python 3.9 for this bundle, I may consider some changes in the CI to support multiple python versions.

yiheng-wang-nv replied:
For the CI error: https://github.com/Project-MONAI/model-zoo/actions/runs/14742610327/job/42137993317?pr=748#step:7:510 Let me push a commit to your branch to fix it. Since detectron2 cannot simply be installed via `pip install detectron2`, we should consider some changes in:
https://github.com/Project-MONAI/model-zoo/blob/dev/ci/get_bundle_requirements.py#L54

yiheng-wang-nv replied:
Since I don't have the permission to push to your forked repo directly, I pushed the changes into: https://github.com/Project-MONAI/model-zoo/tree/730-retinalOCT_RPD_segmentation, could you help to merge it? 
The branch also fixed the pre-commit ci errors.

Then we can see if there are other issues in the dependencies.

ybagdasa replied:
> For the python version, can you confirm if python 3.9 is necessary? I'm not 100% sure, but based on https://detectron2.readthedocs.io/en/latest/tutorials/install.html#requirements , it only requires to use python >= 3.7.
> 
> Currently, our repo uses python 3.10 in default for CI tests. If we do need python 3.9 for this bundle, I may consider some changes in the CI to support multiple python versions.

My attempts to build the detectron2 wheel failed with python 3.10 and it unfortunately seems to be a common issue with their repo: https://github.com/facebookresearch/detectron2/issues/5445

yiheng-wang-nv replied:
Thanks, let me enable py3.9 in the CI

yiheng-wang-nv commented:
Hi @ybagdasa , thanks for the PR contribution. For `pre-commit` CI errors, could you fix them? The auto-push seems cannot work since the branch is pushed from an organization account.

yiheng-wang-nv commented:
Format test can be fixed according to: https://github.com/Project-MONAI/model-zoo/actions/runs/14742610314/job/42137993243?pr=748#step:7:67

yiheng-wang-nv commented:
I will look at this PR later this week

ybagdasa commented:
> Format test can be fixed according to: https://github.com/Project-MONAI/model-zoo/actions/runs/14742610314/job/42137993243?pr=748#step:7:67

@yiheng-wang-nv When running flake8 on the code it flags a lot of mixed-case naming conventions (N800 flags). A lot of the naming comes from the pycocotools api which used mixed-case naming for functions and variables. Having everything lowercase will probably hinder the readability of the code at this point. Is this a strict requirement or can I somehow skip it?

yiheng-wang-nv commented on ci/install_scripts/install_retinalOCT_RPD_segmentation_dependency.sh:
```diff
@@ -0,0 +1,3 @@
+# ensure that using evironment with python==3.9
+pip install -r ./docs/requirements.txt #includes monai
```
We may not need the requirements.txt file in the bundle, it's content is included in the `metadata.json`.
Our CI can generate the requirements file directly via: https://github.com/Project-MONAI/model-zoo/blob/dev/ci/get_bundle_requirements.py#L54

yiheng-wang-nv committed:
fix isort and update cit for detectron2

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv commented:
> > Format test can be fixed according to: https://github.com/Project-MONAI/model-zoo/actions/runs/14742610314/job/42137993243?pr=748#step:7:67
> 
> @yiheng-wang-nv When running flake8 on the code it flags a lot of mixed-case naming conventions (N800 flags). A lot of the naming comes from the pycocotools api which used mixed-case naming for functions and variables. Having everything lowercase will probably hinder the readability of the code at this point. Is this a strict requirement or can I somehow skip it?

Hi @ybagdasa , sorry I cannot see the error you mentioned, the error I saw is from `isort`, could you fix it first? Then we can see the other checks like black and flake8

ybagdasa committed:
Fixed pre-commit and flake8 errors. Added torchvision to metadata.json.

ybagdasa committed:
Merge branch '730-retinalOCT_RPD_segmentation' of github.com:uw-biomedical-ml/monai-model-zoo into 730-retinalOCT_RPD_segmentation

ybagdasa commented:
@yiheng-wang-nv 
I merged your changes with my fixes on my organization repo and gave you write privileges for the repo. I fixed all the formatting issues detected by runtests.sh except for the mixed-case ones from flake8. I also added torchvision to metadata.json since that is required.

yiheng-wang-nv committed:
fix pre-commit and black checks

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv commented:
Hi @ybagdasa , thanks for the updates and providing me the permissions. I pushed a commit to fix the black and pre-commit issues. Now, our CI can reach to the flake8 errors: https://github.com/Project-MONAI/model-zoo/actions/runs/15105923819/job/42454628204?pr=748

I think we can skip N802, N803 and N806 errors, but others may need to be solved

yiheng-wang-nv committed:
skip part of tests and rrevert changes in UNEST bundles

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv commented:
> Hi @ybagdasa , thanks for the updates and providing me the permissions. I pushed a commit to fix the black and pre-commit issues. Now, our CI can reach to the flake8 errors: https://github.com/Project-MONAI/model-zoo/actions/runs/15105923819/job/42454628204?pr=748
> 
> I think we can skip N802, N803 and N806 errors, but others may need to be solved

Hi @ybagdasa , I updated, and could you help to resolve other issues? Like line too long (https://github.com/Project-MONAI/model-zoo/actions/runs/15106735909/job/42456940644?pr=748)

ybagdasa committed:
Fixed flake8 errors. runtests.sh failing at ptype.

ybagdasa commented:
> > Hi @ybagdasa , thanks for the updates and providing me the permissions. I pushed a commit to fix the black and pre-commit issues. Now, our CI can reach to the flake8 errors: https://github.com/Project-MONAI/model-zoo/actions/runs/15105923819/job/42454628204?pr=748
> > I think we can skip N802, N803 and N806 errors, but others may need to be solved
> 
> Hi @ybagdasa , I updated, and could you help to resolve other issues? Like line too long (https://github.com/Project-MONAI/model-zoo/actions/runs/15106735909/job/42456940644?pr=748)

@yiheng-wang-nv I was able to do ./runtests.sh up through pytype. The error popped up here:
`[163/167] check models.maisi_ct_generative.scripts.rectified_flow`

yiheng-wang-nv commented:
> > > Hi @ybagdasa , thanks for the updates and providing me the permissions. I pushed a commit to fix the black and pre-commit issues. Now, our CI can reach to the flake8 errors: https://github.com/Project-MONAI/model-zoo/actions/runs/15105923819/job/42454628204?pr=748
> > > I think we can skip N802, N803 and N806 errors, but others may need to be solved
> > 
> > 
> > Hi @ybagdasa , I updated, and could you help to resolve other issues? Like line too long (https://github.com/Project-MONAI/model-zoo/actions/runs/15106735909/job/42456940644?pr=748)
> 
> @yiheng-wang-nv I was able to do ./runtests.sh up through pytype. The error popped up here: `[163/167] check models.maisi_ct_generative.scripts.rectified_flow`

Hi @ybagdasa , thanks for the updates. The issue does not happen in github's workflow, thus we can ignore it (may due to different environment).

I will help to solve the pre-merge CPU related issues this week, then this PR can be merged.

yiheng-wang-nv committed:
enable different py env for ci tests

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv commented on models/retinalOCT_RPD_segmentation/configs/metadata.json:
```diff
@@ -0,0 +1,145 @@
+        "torchvision": "0.22.0",
+        "detectron2": "0.6",
+        "lxml": "5.4.0",
+        "pillow": "9.5.0"
```
Hi @ybagdasa , I added the pillow requirement due to the issue: https://github.com/facebookresearch/detectron2/issues/5010

yiheng-wang-nv committed:
fix precommit

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv committed:
add conda init for cpu test

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv committed:
update premerge cpu script

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv committed:
add conda init

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv committed:
adjust conda init way

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv committed:
update init conda env

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv commented:
Hi @ybagdasa , I pushed some commits to enable building customized python environment for CI tests, and now `premerge-cpu-check` works.

For the `premerge-gpu-check`, I also updated the script and verified in my local machine. To make it work in Github, there are some internal environment changes (we used internal GPU resources for formal checks). I will do that soon.

yiheng-wang-nv commented:
Hi @ericspod , could you also help to review this new bundle contribution? Thanks in advance!

ericspod commented on setup.cfg:
```diff
@@ -20,7 +20,15 @@ ignore =
+    models/retinalOCT_RPD_segmentation/scripts/analysis_lib.py: N802, N803, N806
+    models/retinalOCT_RPD_segmentation/scripts/datasets/data.py: N802, N803, N806
+    models/retinalOCT_RPD_segmentation/scripts/datasets/volReader.py: N801, N802, N803, N806
+    models/retinalOCT_RPD_segmentation/scripts/inference.py: N802, N803, N806
```
I would suggest fixing the names to conform to PEP8 instead of adding ignores, though if you must you can add ignore comments like `# noqa: N402` in your code rather than here.

ybagdasa replied:
Thanks @ericspod. Just so I understand, is there a portion of the code you are referring to? And what do you see as top-level?

ericspod replied:
I would remove the added lines to the `per_file_ignores` section here and in `analysis_lib.py` for example change names like `Wilson_CI` to `wilson_ci` or something else adhering to the standard, or placing `# noqa: N402` (or other codes for other issues) on those lines to make flake8 ignore them.

As for the top level structures, I mean class and function definitions that are at the top level of files rather than methods or inner functions. It would help understanding what's going on if these have docstring comments at least.

ybagdasa replied:
@ericspod I corrected all the flake8 errors and removed the per_file_ignores entries.

ybagdasa replied:
@ericspod I added docstrings to all the classes and methods.

ericspod replied:
This all looks great, thanks!

ericspod commented on .github/workflows/premerge-cpu.yml:
Why are we using Miniconda for the tests now?

yiheng-wang-nv replied:
Hi @ericspod , this bundle requires python 3.9, but our current CI workflows use python 3.10, we did not meet python version requirements thus did not consider it in CI before. Therefore, I made the changes to enable customized python version support via using conda. What do you think?

ybagdasa replied:
On this issue thread for detectron2 installation with python 3.10, it looks like using setuptools==75.8.0 works for some. Would this be compatible with the current CI workflow?

https://github.com/facebookresearch/detectron2/issues/5453#issuecomment-2918469955

ericspod replied:
I would really recommend doing what's necessary to make this compatible with later versions of Python. Version 3.9 won't be in support for much longer and we should be keeping all bundles running on current platforms. Please do look into the issue with the setuptools dependency and see what else we can do to not depend on that specific version.

ybagdasa replied:
@ericspod I want ahead and added setuptools==75.8.0 to the required packages in metadata.json. I tested the build on python 3.10 and everything seems to install and run. I do get a DEPRECIATION warning from pip now, but I haven't looked into that yet.

ericspod replied:
Please do look into it, but I think that's something we can live with if not fixable currently with the other needed libraries.

ybagdasa replied:
I ran a fresh install and didn't encounter any Depreciation errors this time. I updated the list of packaged in metadata.json to reflect what was installed, including now using monai 1.5. 

@ericspod Will the requirements be installed automatically when the bundle is downloaded? From the user end, how does that work? From my end, how do I simulate the install?

ericspod replied:
For the user the environment has to be setup manually, our testing system creates each environment for each bundle. For now you will have to manually create a virtual environment or conda environment to match the requirements, we need to work on improving that I realise.

ericspod commented on models/retinalOCT_RPD_segmentation/configs/metadata.json:
```diff
@@ -0,0 +1,145 @@
+        "inputs": {
+            "image": {
+                "type": "image",
+                "format": "PNG",
```
"Format" is meant to be about what the pixel values are, eg. just magnitude values or Hounsfield units.

ericspod replied:
This `format` value should be "magnitude" to indicate the values represent differing light or energy levels without units.

ybagdasa replied:
Fixed.

ericspod reviewed (commented):
Hi @ybagdasa and @yiheng-wang-nv I've had a look through the contribution and had a few comments to make. I haven't tried running it but from the conversation it seems it's being tested a lot anyway so I'm sure that's fine. I am concerned that there's a lot of code here and very little actual bundle definition, this code is coming from existing sources so that's understandable as an adaptation into the bundle format. I would still suggest looking at how to reduce this code somehow by using existing libraries perhaps. It's not clear what a lot of the code does either and would benefit with a lot more commentary/docstrings in at least the top level definitions.

I see that Miniconda is being used in the tests now, I think I missed why this was. This required a number of changes to other test scripts so I wonder if it's needed. It might help to provide an environment.yaml file to define what environment to construct rather than rely on commands, this might clean up some of the logic. 

The flake complaints relate to variable names as you see, I would suggest fixing these or using the ignore comments if you don't want to change this code vs. the original source. 

Thanks!

yiheng-wang-nv committed:
Merge branch 'dev' into 730-retinalOCT_RPD_segmentation

ybagdasa committed:
Added docstrings and fixed all flake8 errors.

ybagdasa committed:
Added setuptools==75.8.0 to dependencies to build detectron2 with python 3.10. Modified metadata.json, README.md, and ./docs/requirements.txt files accordingly.

ericspod commented:
Hi @ybagdasa @yiheng-wang-nv this is definitely making progress. If we have a working build for Python 3.10 we can look into removing the Miniconda use in the CICD system and simplify the implementation. Other than that the code is well commented and a few remaining issues can be addressed easily I think.

ybagdasa committed:
Changed "format" in metadata.json to "magnitude"

ybagdasa committed:
Updated required packages in metadata.json.

yiheng-wang-nv committed:
Merge branch 'dev' into 730-retinalOCT_RPD_segmentation

ybagdasa commented:
@ericspod @yiheng-wang-nv I will change the install instructions in the readme to reflect the current working configuration. Other than that, is there anything else left to resolve for this PR?

yiheng-wang-nv commented:
> @ericspod @yiheng-wang-nv I will change the install instructions in the readme to reflect the current working configuration. Other than that, is there anything else left to resolve for this PR?

Hi @ybagdasa , can I confirm now the model supports to work on python 3.10? If so, no other comments on my side. After you push the readme changes, I will revert the conda CI changes and ensure the model works.

ybagdasa committed:
Updated README and bundle requirements.

ybagdasa committed:
Merge branch '730-retinalOCT_RPD_segmentation' of github.com:uw-biomedical-ml/monai-model-zoo into 730-retinalOCT_RPD_segmentation

yiheng-wang-nv committed:
Merge branch 'dev' into 730-retinalOCT_RPD_segmentation

yiheng-wang-nv commented:
Hi @ybagdasa , could you fix the pre commit issue? https://results.pre-commit.ci/run/github/490124346/1753149419.xfcqB10AQdC5-UPubS5znQ

ybagdasa committed:
Fixed pre-commit eof line.

ybagdasa committed:
Merge branch '730-retinalOCT_RPD_segmentation' of github.com:uw-biomedical-ml/monai-model-zoo into 730-retinalOCT_RPD_segmentation

ybagdasa commented:
> Hi @ybagdasa , could you fix the pre commit issue? https://results.pre-commit.ci/run/github/490124346/1753149419.xfcqB10AQdC5-UPubS5znQ

Done :)

ybagdasa commented:
@yiheng-wang-nv @ericspod Is there anything else I need to do for this?

ericspod commented:
Hi @ybagdasa if you've addressed things then please turn this into a full PR (not a draft) and @yiheng-wang-nv can have one more look at things. I'll enable tests and see how they do as well.

yiheng-wang-nv committed:
revert py39 changes

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv committed:
revert py39 changes

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv commented:
/build

yiheng-wang-nv commented:
Hi @ybagdasa I reverted the py3.9 related changes, this PR can merge after passing all py3.10 based tests

yiheng-wang-nv committed:
add customized detectron2

Signed-off-by: Yiheng Wang <vennw@nvidia.com>

yiheng-wang-nv commented:
/build

yiheng-wang-nv commented:
Hi @ybagdasa @ericspod , now this bundle works fine with python 3.10, I will merge the PR. Thanks!
